gufe/transformations/transformation.py

```diff
@@ -0,0 +1,263 @@
+        return self._protocol
+
+    @property
+    def mapping(self):
+        """The mapping between atoms in `initial` to `final` `ChemicalSystem`s for
+        the transformation.
+
+        """
+        return self._mapping
+
+    @property
+    def name(self):
+        """User-specified for the transformation; used as part of its hash."""
+        return self._name
+
+    def __lt__(self, other):
+        return hash(self) < hash(other)
+
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            return False
+        if self._name != other.name:
+            return False
+        if self._protocol != other.protocol:
+            return False
+        if self._mapping != other.mapping:
+            return False
+        if self._final != other.final:
+            return False
+        if self._initial != other.initial:
+            return False
+
+        return True
+
+    def __hash__(self):
+        return hash(
+            (
+                self._initial,
+                self._final,
+                self._mapping,
+                self._name,
+                self._protocol,
+            )
+        )
+
+    # TODO: broken without a `Protocol` registry of some kind
+    # should then be changed to use the registry
+    def to_dict(self) -> dict:
+        return {
+            "initial": self.initial.to_dict(),
+            "final": self.final.to_dict(),
+            "protocol": self.protocol.to_dict(),
+            "mapping": self.mapping.to_dict(),
+        }
+
+    # TODO: broken without a `Protocol` registry of some kind
+    # should then be changed to use the registry
+    @classmethod
+    def from_dict(cls, d: dict):
+        return cls(
+            initial=ChemicalSystem.from_dict(d["initial"]),
+            final=ChemicalSystem.from_dict(d["final"]),
+            protocol=Protocol.from_dict(d["protocol"]),
+            mapping=Mapping.from_dict(d["mapping"]),
+        )
+
+    def create(self) -> ProtocolDAG:
+        """Returns a `ProtocolDAG` executing this `Transformation.protocol`."""
+        return self.protocol.create(
+            initial=self.initial,
+            final=self.final,
+            mapping=self.mapping,
+            name=str(self.__hash__()),
+        )
+
+    def gather(
+        self, protocol_dag_results: Iterable[ProtocolDAGResult]
+    ) -> ProtocolResult:
+        """Gather multiple `ProtocolDAGResult`s into a single `ProtocolResult`.
+
+        Parameters
+        ----------
+        protocol_dag_results : Iterable[ProtocolDAGResult]
+            The `ProtocolDAGResult`s to assemble aggregate quantities from.
+
+        Returns
+        -------
+        ProtocolResult
+            Aggregated results from many `ProtocolDAGResult`s from a given `Protocol`.
+
+        """
+        return self.protocol.gather(protocol_dag_results=protocol_dag_results)
+
+
+# we subclass `Transformation` here for typing simplicity
+class NonTransformation(Transformation):
```

Comment:
So throwing my 2 cent here, I don't really like the `Transformation` and `NonTransformation` naming (especially the latter), it's strangely non-intuitive (especially if you have nontransformation sitting in transformations, I honestly can just see users sitting there being confused).

Is there an alternative we could use here? The concept of AlchemicalPaths (through states) or just calling them Experiments? Even being more descriptive with AlchemicalTransform(ation) would probably ease some of the likely pain points users are going to feel here.

Reply:
Experiment sounds like a bench thing.

What a `Transformation` is the application of a `Protocol` onto specific chemical systems.  It is also used as the edge on alchemical Graphs.

You could call it `ParametrisedProtocol` but I feel like we've already got too many names revolving around Protocol.

I don't hate `AlchemicalTransformation`

Reply:
Agree `NonTransformation` is a bit obtuse. I think it's also totally descriptive without stacking in any additional, unintended meaning. One could even imagine it as a lambda protocol that changes a system to itself, effectively just doing dynamics and no alchemistry.

I'm fine with `Transformation` being renamed to `AlchemicalTransformation`, which goes with our `AlchemicalNetwork`.

Reply:
I'll kick this down the road to the "great renaming" event we can have once we've got all the objects.